File: lib/src/main/java/io/github/kory33/guardedqueries/core/subqueryentailments/LocalInstanceTerm.java
package io.github.kory33.guardedqueries.core.subqueryentailments;

import uk.ac.ox.cs.pdq.fol.Constant;
import uk.ac.ox.cs.pdq.fol.Term;
import uk.ac.ox.cs.pdq.fol.Variable;

import java.util.function.Function;
import java.util.function.Predicate;

public sealed interface LocalInstanceTerm {
    record LocalName(int value) implements LocalInstanceTerm {
        @Override
        public Term mapLocalNamesToTerm(Function<? super LocalName, ? extends Term> mapper) {
            return mapper.apply(this);
        }

        @Override
        public String toString() {
            return "LocalName[" + value + "]";
        }
    }

    record RuleConstant(Constant constant) implements LocalInstanceTerm {
        @Override
        public Term mapLocalNamesToTerm(Function<? super LocalName, ? extends Term> mapper) {
            return this.constant;
        }

        @Override
        public String toString() {
            return "RuleConstant[" + constant + "]";
        }
    }

    default boolean isConstantOrSatisfies(final Predicate<? super LocalName> predicate) {
        if (this instanceof RuleConstant) {
            return true;
        } else if (this instanceof LocalName localName) {
            return predicate.test(localName);
        } else {
            throw new IllegalStateException("Unreachable: " + this);
        }
    }

    Term mapLocalNamesToTerm(Function<? super LocalName, ? extends Term> mapper);

    static LocalInstanceTerm fromTermWithVariableMap(
            final Term term,
            final Function<? super Variable, ? extends LocalInstanceTerm> mapper
    ) {
        if (term instanceof Constant constant) {
            return new RuleConstant(constant);
        } else if (term instanceof Variable variable) {
            return mapper.apply(variable);
        } else {
            throw new IllegalArgumentException("Unsupported term: " + term);
        }
    }
}

File: lib/src/main/scala/io/github/kory33/guardedqueries/core/subqueryentailments/LocalInstanceTerm.scala
package io.github.kory33.guardedqueries.core.subqueryentailments

import uk.ac.ox.cs.pdq.fol.Constant
import uk.ac.ox.cs.pdq.fol.Term
import uk.ac.ox.cs.pdq.fol.Variable
import java.util.function.Function
import java.util.function.Predicate

sealed trait LocalInstanceTerm {
  def isConstantOrSatisfies(predicate: Predicate[_ >: LocalInstanceTerm.LocalName]): Boolean =
    this match
      case _: LocalInstanceTerm.RuleConstant      => true
      case localName: LocalInstanceTerm.LocalName => predicate.test(localName)

  def mapLocalNamesToTerm(mapper: Function[_ >: LocalInstanceTerm.LocalName, _ <: Term]): Term
}

object LocalInstanceTerm {
  case class LocalName(value: Int) extends LocalInstanceTerm {
    override def mapLocalNamesToTerm(mapper: Function[
      _ >: LocalInstanceTerm.LocalName,
      _ <: Term
    ]): Term = mapper.apply(this)

    override def toString: String = "LocalName[" + value + "]"
  }

  case class RuleConstant(constant: Constant) extends LocalInstanceTerm {
    override def mapLocalNamesToTerm(mapper: Function[
      _ >: LocalInstanceTerm.LocalName,
      _ <: Term
    ]): Term = this.constant
    override def toString: String = "RuleConstant[" + constant + "]"
  }

  def fromTermWithVariableMap(term: Term,
                              mapper: Function[_ >: Variable, _ <: LocalInstanceTerm]
  ): LocalInstanceTerm =
    term match
      case constant: Constant => RuleConstant(constant)
      case variable: Variable => mapper.apply(variable)
      case _                  => throw new IllegalArgumentException("Unsupported term: " + term)
}
